### PR TITLE
fpclassify wrapper

### PR DESCRIPTION
**Pull request by daschuer**
This PR adds an fpclassify wrapper as discussed here: 
https://github.com/mixxxdj/mixxx/pull/803
Which was also merges here. 

I can't make the denormals test work .. strange.

### CODE REVIEW

File: src/util/fpclassify.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
+// this is a wapper around the fpclassify functions which prevents inlining 
+// It is compiled without optization 
+// The rest of the source of Mixxx is compiled with -ffast-math, which braks
```
Comment (ywwg) [inline src/util/fpclassify.cpp]: breaks

**Comment by daschuer** [comment]
Test is working again. We needed to use volatile to stop the complier from optimize out test away.

File: src/util/math.h [hunk 2]
```diff
@@ -34,22 +35,6 @@ inline bool even(T value) {
     return value % 2 == 0;
 }
 
-#ifdef _MSC_VER
-// VC++ uses _isnan() instead of isnan() and !_finite instead of isinf.
-#include <float.h>
-#define isnan(x) _isnan(x)
-#define isinf(x) (!_finite(x))
-// Ask VC++ to emit an intrinsic for fabs instead of calling std::fabs.
-#pragma intrinsic(fabs)
```
Comment (rryan) [inline src/util/math.h]: Looks like you dropped this one.

File: src/util/fpclassify.h [hunk 3]
```diff
@@ -0,0 +1,17 @@
+
```
Comment (rryan) [inline src/util/fpclassify.h]: please add an include guard

**Comment by rryan** [comment]
Hm, I think this drops important other CCFLAGS like ```-isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -mmacosx-version-min=10.9``` which determines the SDK used on OS X.

**Comment by rryan** [comment]
Here are the two different invocations:
```
clang++ -o osx64_build/test/mathutiltest.o -c -isystem /usr/local/homebrew/include -isystem /usr/local/qt4/include/ -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -mmacosx-version-min=10.9 -arch x86_64 -pipe -Wall -Wextra -g -F/usr/local/qt4/lib -O3 -ffast-math -funroll-loops -fomit-frame-pointer -march=native -fsanitize=address -fno-omit-frame-pointer -Dx86_64 -DMIXXX_BUILD_DEBUG -DSETTINGS_FILE=\"mixxx.cfg\" -D__PORTAUDIO__ -DQT_SHARED -DQT_TABLET_SUPPORT -DQT_CORE_LIB -DQT_GUI_LIB -DQT_OPENGL_LIB -DQT_XML_LIB -DQT_SVG_LIB -DQT_SQL_LIB -DQT_SCRIPT_LIB -DQT_NETWORK_LIB -D__SNDFILE__ -D__COREAUDIO__ -D__HID__ -D__VINYLCONTROL__ -D__SHOUTCAST__ -D__AUTODJCRATES__ -I/usr/local/include -Iosx64_build -Isrc -Ilib/soundtouch-1.8.0 -Ilib/replaygain -I/usr/local/qt4/lib/QtCore.framework/Headers -I/usr/local/qt4/lib/QtGui.framework/Headers -I/usr/local/qt4/lib/QtOpenGL.framework/Headers -I/usr/local/qt4/lib/QtXml.framework/Headers -I/usr/local/qt4/lib/QtSvg.framework/Headers -I/usr/local/qt4/lib/QtSql.framework/Headers -I/usr/local/qt4/lib/QtScript.framework/Headers -I/usr/local/qt4/lib/QtXmlPatterns.framework/Headers -I/usr/local/qt4/lib/QtNetwork.framework/Headers -I/usr/local/qt4/lib/QtTest.framework/Headers -I/usr/local/qt4/lib/QtScriptTools.framework/Headers -Ilib/gtest-1.7.0/include -Ilib/fidlib-0.9.10 -I/usr/include/taglib -I/System/Library/Frameworks/Security.framework/Headers -I/System/Library/Frameworks/CoreServices.framework/Headers -Ilib/apple -Ilib/hidapi-0.8.0-rc1/hidapi -Ilib/xwax -Ilib/scratchlib -Ilib/vamp-2.3 -Ilib/gtest-1.7.0/include -Ilib/gmock-1.7.0/include src/test/mathutiltest.cpp
clang++ -o osx64_build/util/fpclassify.o -c -isystem /usr/local/homebrew/include -isystem /usr/local/qt4/include/ -Dx86_64 -DMIXXX_BUILD_DEBUG -DSETTINGS_FILE=\"mixxx.cfg\" -D__PORTAUDIO__ -DQT_SHARED -DQT_TABLET_SUPPORT -DQT_CORE_LIB -DQT_GUI_LIB -DQT_OPENGL_LIB -DQT_XML_LIB -DQT_SVG_LIB -DQT_SQL_LIB -DQT_SCRIPT_LIB -DQT_NETWORK_LIB -D__SNDFILE__ -D__COREAUDIO__ -D__HID__ -D__VINYLCONTROL__ -D__SHOUTCAST__ -D__AUTODJCRATES__ -I/usr/local/include -Iosx64_build -Isrc -Ilib/soundtouch-1.8.0 -Ilib/replaygain -I/usr/local/qt4/lib/QtCore.framework/Headers -I/usr/local/qt4/lib/QtGui.framework/Headers -I/usr/local/qt4/lib/QtOpenGL.framework/Headers -I/usr/local/qt4/lib/QtXml.framework/Headers -I/usr/local/qt4/lib/QtSvg.framework/Headers -I/usr/local/qt4/lib/QtSql.framework/Headers -I/usr/local/qt4/lib/QtScript.framework/Headers -I/usr/local/qt4/lib/QtXmlPatterns.framework/Headers -I/usr/local/qt4/lib/QtNetwork.framework/Headers -I/usr/local/qt4/lib/QtTest.framework/Headers -I/usr/local/qt4/lib/QtScriptTools.framework/Headers -Ilib/gtest-1.7.0/include -Ilib/fidlib-0.9.10 -I/usr/include/taglib -I/System/Library/Frameworks/Security.framework/Headers -I/System/Library/Frameworks/CoreServices.framework/Headers -Ilib/apple -Ilib/hidapi-0.8.0-rc1/hidapi -Ilib/xwax -Ilib/scratchlib -Ilib/vamp-2.3 src/util/fpclassify.cpp
```

**Comment by rryan** [comment]
This is a more targeted way to remove it that leaves in the other flags:
```
diff --git a/build/depends.py b/build/depends.py
index 98d92ff..611622d 100644
--- a/build/depends.py
+++ b/build/depends.py
@@ -494,12 +494,13 @@ class ProtoBuf(Dependence):

 class FpClassify(Dependence):
     # This is a wrapper arround the fpclassify function that pevents inlining
     # It is compiled without optimization and allows to use these function
     # from -ffast-math optimized objects
     def sources(self, build):
-        # add this file without optimization
-        env=build.env.Clone()
-        return env.Object('util/fpclassify.cpp', CCFLAGS='')
+        # add this file without optimization
+        env = build.env.Clone()
+        env['CCFLAGS'].remove('-ffast-math')
+        return env.Object('util/fpclassify.cpp')


 class MixxxCore(Feature):
```

**Comment by daschuer** [comment]
Notes addressed

**Comment by rryan** [comment]
Thanks for fix! The tests pass on my Linux machine (while they didn't in #803) and on my OS X machine (which passed in both #803 and here). Should we get a Windows test in before merging?

**Comment by rryan** [comment]
Hm I wonder if the MSVC optimization ```/fp:fast``` has the same behavior w.r.t. ```isnan``` as ```-ffast-math```? In which case we should probably remove both of them when building fpclassify.cpp?

**Comment by daschuer** [comment]
Ah yes. In any case ,it should be save to remove the flag as well.
We should run the test on windows, but luckily jenkins does it for us after merge.

**Comment by sblaisot** [comment]
I can provide some help with that and test it on my win7/64 machine, but I need some detailed instructions on what you need me to do.

**Comment by daschuer** [comment]
Cool, Thank you. 
* compile this with "scons test=1"
* run the test "mixxx-test.exe --gtest_filter=MathUtilTest*"

If the test fails, we need to remove the /fp:fast flag as well

**Comment by sblaisot** [comment]
Configuring FFMPEG
ValueError: list.remove(x): x not in list:
  File "D:\Documents\SÚbastien\Mes Documents\projets\mixxx\SConstruct", line 73:

    SConscript(File('src/SConscript.env'), variant_dir=Dir(build.build_dir), duplicate=0)
  File "C:\Python27\Scripts\..\Lib\site-packages\scons-2.3.3\SCons\Script\SConscript.py", line 613:
    return method(*args, **kw)
  File "C:\Python27\Scripts\..\Lib\site-packages\scons-2.3.3\SCons\Script\SConscript.py", line 550:
    return _SConscript(self.fs, *files, **subst_kw)
  File "C:\Python27\Scripts\..\Lib\site-packages\scons-2.3.3\SCons\Script\SConscript.py", line 260:
    exec _file_ in call_stack[-1].globals
  File "D:\Documents\SÚbastien\Mes Documents\projets\mixxx\src\SConscript.env", line 96:
    dep_sources = dependency.sources(build)
  File "D:\Documents\SÚbastien\Mes Documents\projets\mixxx\build\depends.py", line 502:
    env['CCFLAGS'].remove('-ffast-math')
  File "C:\Python27\lib\UserList.py", line 79:
    def remove(self, item): self.data.remove(item)

**Comment by sblaisot** [comment]
just commented out `env['CCFLAGS'].remove('-ffast-math')` and compiling ok. I'll post test results
and I'll make a second try replacing this line with `env['CCFLAGS'].remove('fp:fast')` and post test results too
we will need to add 
```
if build.platform_is_windows:
            if build.toolchain_is_msvs:
                        env['CCFLAGS'].remove('fp:fast')
            else:
                        env['CCFLAGS'].remove('-ffast-math')
```

**Comment by sblaisot** [comment]
ok, first try with said line commented out, here are the test results:
```
<?xml version="1.0" encoding="UTF-8"?>
<testsuites tests="4" failures="0" disabled="0" errors="0" timestamp="2015-12-13T10:50:54" time="0.003" name="AllTests">
  <testsuite name="MathUtilTest" tests="4" failures="0" disabled="0" errors="0" time="0.001">
    <testcase name="MathClampUnsafe" status="run" time="0" classname="MathUtilTest" />
    <testcase name="IsNaN" status="run" time="0.001" classname="MathUtilTest" />
    <testcase name="IsInf" status="run" time="0" classname="MathUtilTest" />
    <testcase name="Denormal" status="run" time="0" classname="MathUtilTest" />
  </testsuite>
</testsuites>
```

**Comment by sblaisot** [comment]
with previously faulty line changed to `env['CCFLAGS'].remove('/fp:fast')` (note that you have to specify the / on the windows platform) here are the test results:
```
<?xml version="1.0" encoding="UTF-8"?>
<testsuites tests="4" failures="0" disabled="0" errors="0" timestamp="2015-12-13T10:59:57" time="0.001" name="AllTests">
  <testsuite name="MathUtilTest" tests="4" failures="0" disabled="0" errors="0" time="0">
    <testcase name="MathClampUnsafe" status="run" time="0" classname="MathUtilTest" />
    <testcase name="IsNaN" status="run" time="0" classname="MathUtilTest" />
    <testcase name="IsInf" status="run" time="0" classname="MathUtilTest" />
    <testcase name="Denormal" status="run" time="0" classname="MathUtilTest" />
  </testsuite>
</testsuites>
```

**Comment by daschuer** [comment]
Thank you. That means, that it is not required to remove /fp:fast right? 
Cool. That means that we can turn back to the original solution for Windows.

**Comment by daschuer** [comment]
@sblaisot: based on your results, I have added a commit, that should work on windows. 
Can you verify it? Thank you.

**Comment by sblaisot** [comment]
1_ yes, it compiles without any problem
2_ tests results are the same, no problem

I only tested 64bits compilation.

**Comment by sblaisot** [comment]
please not: with this branch, playing all tests, I have 10 tests out of 35 that fail on softtakeover. I don't know if this is expected or not.

**Comment by daschuer** [comment]
Thank you!
The softtakover tests are an other story. 
@rryan: I think we can merge, right?

**Comment by rryan** [comment]
> ValueError: list.remove(x): x not in list:

LGTM @daschuer, thank you -- though it looks like maybe we should test for presence of hte optimization flag? Will this throw an exception with optimize=off?

**Comment by rryan** [comment]
oh nvm, you added that already

**Pull request merged**